src/main/java/seedu/address/logic/commands/ChangeCommand.java

```diff
@@ -0,0 +1,208 @@
+        requireNonNull(model);
+        List<Appointment> lastShownList = model.getFilteredAppointmentList();
+
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_APPOINTMENT_DISPLAYED_INDEX);
+        }
+
+        Appointment appointmentToEdit = lastShownList.get(index.getZeroBased());
+        Appointment editedAppointment = createEditedAppointment(appointmentToEdit, editAppointmentDescriptor);
+
+        if (appointmentToEdit.equals(editedAppointment) && model.hasAppointment(editedAppointment)) {
+            throw new CommandException(MESSAGE_DUPLICATE_APPOINTMENT);
+        }
+        model.deleteAppointment(appointmentToEdit);
+        if (model.hasOverlappingAppointment(editedAppointment)) {
+            model.addAppointment(appointmentToEdit);
+            throw new CommandException(ASSIGNMENT_OVERLAP);
+        }
+        model.addAppointment(appointmentToEdit);
+        model.setAppointment(appointmentToEdit, editedAppointment);
+        model.updateFilteredAppointmentList(PREDICATE_SHOW_ALL_APPOINTMENTS);
+        return new CommandResult(String.format(MESSAGE_EDIT_APPOINTMENT_SUCCESS, editedAppointment));
+    }
+
+    /**
+     * Creates and returns an {@code Appointment} with the details of {@code appointmentToEdit}
+     * edited with {@code editAppointmentDescriptor}.
+     */
+    private static Appointment createEditedAppointment(Appointment appointmentToEdit,
+                                                       EditAppointmentDescriptor editAppointmentDescriptor) {
+        assert appointmentToEdit != null;
+        assert editAppointmentDescriptor != null;
+        assert editAppointmentDescriptor.duration != null;
+        assert editAppointmentDescriptor.date != null;
+        assert editAppointmentDescriptor.startTime != null;
+
+        Date date = editAppointmentDescriptor.getDate().get();
+        Time startTime = editAppointmentDescriptor.getStartTime().get();
+        Duration duration = editAppointmentDescriptor.getDuration().get();
+        Time endTime = new Time(startTime.getTime().plus(duration));
+
+        Patient patient = appointmentToEdit.getPatient();
+        return new Appointment(date, startTime, endTime, patient);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        // short circuit if same object
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof ChangeCommand)) {
+            return false;
+        }
+
+        // state check
+        ChangeCommand e = (ChangeCommand) other;
+        return index.equals(e.index)
+                && editAppointmentDescriptor.equals(e.editAppointmentDescriptor);
+    }
+
+    /**
+     * Stores the details to edit the appointment with. Each field value will replace the
+     * corresponding field value of the appointment.
+     */
+    public static class EditAppointmentDescriptor {
```

Comment:
🤣 
How about we take out our code of the `EditAppointmentDescriptor` and `DurationSupporter` to have a common class!